### ISSUE TITLE
Offscreen renderer compositor in gpu mode fails and falls back to SoftwareOutputDevice

### ISSUE THREAD

**Issue by vulture** [electron/electron#11439]
When using the offscreen renderer WITHOUT disabling hardware acceleration, it seems like the compositor output device fails to open a GPU channel and (after 5 tries) falls back to using the SoftwareOutputDevice. Seems like this doesn't happen for non-OSR windows (similar code path succeeds in the compositor code, when I set a breakpoint around the failure point), so it may be something specific to OSR.

electron --enable-logging --v=1 main.js yields 5x :
`[31672:1214/093050.323:ERROR:gl_surface_egl.cc(785)] eglCreateWindowSurface failed with error EGL_BAD_ALLOC`

More details following the template data.

* Electron version: 1.8.2beta3, 1.7.9, 1.7.3, 1.4.15
* Operating system: Win 8.1, Win 10

### Expected behavior

GPU channel for compositor is created.

### Actual behavior

software_output_device_ is created instead.

### How to reproduce

1. Grab the OSR example from the electron github: https://github.com/electron/electron/blob/master/docs/tutorial/offscreen-rendering.md
2. Comment out the `app.disableHardwareAcceleration()` line.
3. Run `electron --enable-logging --v=1 main.js`  (or whatever js you copy that example to)
4. Observe `[31672:1214/093050.323:ERROR:gl_surface_egl.cc(785)] eglCreateWindowSurface failed with error EGL_BAD_ALLOC` in the log.

More very verbose details:

I am hoping to use the OSR composited GPU texture directly on the Electron side, staying on the GPU, since we'll be doing other things with this texture on the GPU (without pulling down the pixels to system memory, then re-uploading it to GPU to re-render with it, which is slow). This also allows us to take advantage of e.g. WebGL, etc.

From my understanding, in the main process:
`libcc : src\content\browser\compositor\gpu_process_transport_factory.cc`
`GpuProcessTransportFactory::EstablishedGpuChannel` calls some GPU channel IPC code in `libcc : src\gpu\ipc\client\command_buffer_proxy_impl.cc` and &result returns false (or doesn't set true?), which causes an IPC failure: `Failed to send GpuChannelMsg_CreateCommandBuffer`. This results in `gpu_process_transport_factory.cc` `bool created_gpu_browser_compositor = !!context_provider && !!shared_worker_context_provider_;` to be false. So, the GPU channel setup has failed for the compositor. It tries up to 5 times:
```
    if (!created_gpu_browser_compositor) {
      // Try again.
      gpu::GpuChannelEstablishedCallback callback(
          base::Bind(&GpuProcessTransportFactory::EstablishedGpuChannel,
                     callback_factory_.GetWeakPtr(), compositor,
                     create_gpu_output_surface, num_attempts + 1));
      DCHECK(gpu_channel_factory_);
      gpu_channel_factory_->EstablishGpuChannel(callback);
      return;
    }
```
but this fails for the same reason every time.

Looking at the gpu process, presumably in response to the IPC message sent from the main process, it's the original error message from the verbose logging:
`[31672:1214/093050.323:ERROR:gl_surface_egl.cc(785)] eglCreateWindowSurface failed with error EGL_BAD_ALLOC`
`libcc : src\ui\gl\gl_surface_egl.cc` inside `NativeViewGLSurfaceEGL::Initialize` calls
```
  // Create a surface for the native window.
  surface_ = eglCreateWindowSurface(
      GetDisplay(), GetConfig(), window_, &egl_window_attributes[0]);

  if (!surface_) {
    LOG(ERROR) << "eglCreateWindowSurface failed with error "
               << GetLastEGLErrorString();
```
which fails. So, maybe the way Electron sets up the OSR is incompatible with eglCreateWindowSurface? The error message is unfortunately not very specific.

After these 5 failures, it falls back to using a software compositor output:
```
    if (!create_gpu_output_surface) {
      display_output_surface =
          base::MakeUnique<SoftwareBrowserCompositorOutputSurface>(
              CreateSoftwareOutputDevice(compositor.get()), vsync_callback,
              compositor->task_runner());
```
which calls back to Electron's `OffScreenRenderWidgetHostView::CreateSoftwareOutputDevice`, creates a `software_output_device_ = new OffScreenOutputDevice(...)`, and once this is set, `OffScreenRenderWidgetHostView::OnSwapCompositorFrame` runs:
```
    if (software_output_device_) {
      ... cpu path ...
    } else {
      ... gpu path? ...
    }
```
and the gpu path essentially never gets used. So we aren't able to get a GPU texture of the compositor output in the main process to use for other things.

Sorry in advance for such a complex problem, maybe masked in the past due to the built-in fallback code, but maybe it's some simple setting that would make more sense to you guys. We were really hoping to be able to use GPU acceleration in the compositor output if possible! Thanks!

**Comment by vulture** [comment]
Found out the libGLESv2.dll symbols were in there, just not working right for some reason, so was able to dig in a little more:
`libcc : src\ui\gl\gl_surface_egl.cc`
```
  surface_ = eglCreateWindowSurface(
      GetDisplay(), GetConfig(), window_, &egl_window_attributes[0]);

  if (!surface_) {
    LOG(ERROR) << "eglCreateWindowSurface failed with error "
               << GetLastEGLErrorString();
    Destroy();
    return false;
```
calls `libcc : src\third_party\angle\src\libGLESv2\entry_points_egl.cpp`
```
    Error error = ValidateCreateWindowSurface(display, configuration, win, attributes);
```
calls `libcc : src\third_party\angle\src\libANGLE\validationEGL.cpp`
```
    if (Display::hasExistingWindowSurface(window))
    {
        return Error(EGL_BAD_ALLOC);
    }
```

So, it appears as though the window already has a GPU surface created for it, and 1 is the limit.

**Comment by brenca** [comment]
Sorry for the slow response, but I've tested this on my machine with v1.8.2-beta.3, and I don't see the error on the log. I'm positive that the GPU render path is working on my machine, so this might be machine dependent.

**Comment by vulture** [comment]
We tested on 1 laptop with dedicated gpu and gpu bindings appropriately set (repro'd), 1 desktop with dedicated gpu (repro'd). However, we just tested on another different laptop with dedicated gpu (not sure about gpu bindings) and ran into an interesting "issue" -- if we ran with hardware accel disabled in the js, then we got the white window about "No content under offscreen mode". If we commented out the app.disable... call (i.e., run in gpu mode), then instead of this white window, it actually showed the webpage content (e.g. http://github.com). On our other machines, we get this white window. Does that happen for you? Could that alter the behavior of the OSR/GPU pipelines?

**Comment by limoxi** [comment]
![snipaste_2018-02-02_19-36-48](https://user-images.githubusercontent.com/7345689/35731419-6ed4f1a6-0850-11e8-8aee-6b338340d6ad.png)
I got this too.

**Comment by brenca** [comment]
The `No content under offscreen mode` window is the intended behavior. It should behave consistently now in both GPU and CPU rendering modes.

If this problem still happens for you, let me know.

**Comment by brenca** [comment]
BTW this could very well be an upstream bug, I'm not sure. Let's see if the newer electron versions improved this.

I would also be interested in how you use the texture in the same process (presumable in a native addon?), that feature - if you want to - could be pulled into electron itself.

**Comment by vulture** [comment]
Thanks for the update. We're in the process of making our code work with Electron 3+, so once that's done, I will be swinging back to this and seeing what the state of the issue is.

@brenca -- sorry for the (very) late reply -- we're just modding Electron's OSR C++ code at this point. They are pretty hefty and hacky modifications (in an attempt to maximize performance) and probably not something Electron would want to pull back in (though if Electron had a friendlier way of doing it, we'd probably go that route instead of rolling our own!)

**Comment by sofianguy** [comment]
@vulture I'm going to close this as we are no longer releasing bug fixes for versions earlier than 2.0.x. Feel free to reach out if you are seeing the bug on versions later than 2.0.x.
